### PR TITLE
rather use format_string for checking

### PR DESCRIPTION
**Pull request by awalterschulze**
Changed --check to rather use format_string, this way stdin is also supported and API is simpler.

## Test Plan

```sh
$ rebar3 as release escriptize

$ _build/release/bin/erlfmt --check "{src,include,test}/*.{hrl,erl}" "rebar.config"
Checking formatting...
[warn] src/erlfmt_cli.erl
[warn] src/erlfmt_scan.erl
[warn] test/erlfmt_SUITE.erl
[warn] Code style issues found in the above file(s). Forgot to run erlfmt?

$ _build/release/bin/erlfmt -w "{src,include,test}/*.{hrl,erl}" "rebar.config"
$ git diff
...
modified:   src/erlfmt_cli.erl
modified:   src/erlfmt_scan.erl
modified:   test/erlfmt_SUITE.erl
...

$ _build/release/bin/erlfmt --check "{src,include,test}/*.{hrl,erl}" "rebar.config"
Checking formatting...
All matched files use erlfmt code style!
```

### CODE REVIEW

File: src/erlfmt_cli.erl [hunk 1]
```diff
@@ -83,18 +93,48 @@ do_unprotected(Opts, Name) ->
             erlang:halt(2)
     end.
 
+-dialyzer({no_improper_lists, [read_stdin/1]}).
+
+read_stdin(Data) ->
+    case io:get_chars(standard_io, "", 4096) of
+        MoreData when is_binary(MoreData) -> read_stdin([Data | MoreData]);
+        eof -> {ok, Data};
+        {error, Reason} -> throw({error, Reason})
```
Comment (michalmuskala) [inline src/erlfmt_cli.erl:102]: Do we handle throws nicely in this context? We handle them in `erlfmt`, but I'm not sure here

File: src/erlfmt_cli.erl [hunk 2]
```diff
@@ -83,18 +93,48 @@ do_unprotected(Opts, Name) ->
...
+                Original = unicode:characters_to_list(OriginalBin),
+                case erlfmt:format_string(Original, Options) of
+                    {ok, Formatted, FormatWarnings} ->
+                        case Formatted of
+                            Original -> {ok, FormatWarnings};
+                            _ -> {warn, FormatWarnings}
+                        end;
+                    Other ->
+                        Other
+                end;
```
Comment (michalmuskala) [inline src/erlfmt_cli.erl:129]: Let's extract this to a function

File: src/erlfmt_cli.erl [hunk 3]
```diff
@@ -139,10 +179,8 @@ parse_opts([insert_pragma | Rest], Name, Files, Config) ->
     parse_opts(Rest, Name, Files, Config#config{pragma = insert});
 parse_opts([{files, NewFiles} | Rest], Name, Files0, Config) ->
     parse_opts(Rest, Name, expand_files(NewFiles, Files0), Config);
-parse_opts([], _Name, [stdin], #config{out = Out}) when Out =/= standard_out ->
-    {error, "stdin mode can't be combined with out or write options"};
-parse_opts([], _Name, [stdin], #config{out = check}) ->
-    {error, "stdin mode can't be combined check mode"};
+parse_opts([], _Name, [stdin], #config{out = Out}) when Out =/= standard_out, Out =/= check ->
```
Comment (michalmuskala) [inline src/erlfmt_cli.erl:182]: `out` can't be `check` now, right?

File: src/erlfmt_scan.erl [hunk 4]
```diff
@@ -136,6 +136,8 @@ continue(Scan, Inner0, Loc0, Buffer0) ->
 read_rest(#state{inner = undefined}) ->
     %% reached EOF, no further nodes
     {ok, ""};
+read_rest(#state{scan = _Scan, inner = eof, loc = _Loc, buffer = Buffer}) ->
```
Comment (michalmuskala) [inline src/erlfmt_scan.erl:139]: When does this happen?

File: src/erlfmt_cli.erl (hunk 2 above)
Comment (michalmuskala) [reply src/erlfmt_cli.erl:129]: Also, does this compose well with `require-pragma`?

File: src/erlfmt_cli.erl [hunk 5]
```diff
@@ -30,10 +31,10 @@ opts() ->
         {write, $w, "write", undefined, "modify formatted files in place"},
         {out, $o, "out", binary, "output directory"},
         {verbose, undefined, "verbose", undefined, "include debug output"},
-        {check, $c, "check", undefined,
+        {check, undefined, "check", undefined,
```
Comment (michalmuskala) [inline src/erlfmt_cli.erl:34]: Why this change?

File: src/erlfmt_cli.erl (hunk 1 above)
Comment (awalterschulze) [reply src/erlfmt_cli.erl]: do calls and catch throws from do_unprotected which calls format_file, so I think we are fine.
write_formatted also contained throws which we didn't try to catch here.

File: src/erlfmt_cli.erl (hunk 2 above)
Comment (awalterschulze) [reply src/erlfmt_cli.erl]: Great idea, I extracted it.
I also added a test for require-pragma

File: src/erlfmt_scan.erl (hunk 4 above)
Comment (awalterschulze) [reply src/erlfmt_scan.erl:139]: This was for no_pragma.erl, which has only one line.
We read a single form to check for pragma, so when we call read_rest, the only line has already been read.

File: src/erlfmt_cli.erl (hunk 5 above)
Comment (awalterschulze) [reply src/erlfmt_cli.erl:34]: I don't know, it just seems more symmetric with --verbose

File: src/erlfmt_cli.erl (hunk 3 above)
Comment (awalterschulze) [reply src/erlfmt_cli.erl]: Good catch. Fixed :)

**Comment by awalterschulze** [comment]
This newest commit changes the API and factors out the writing of the file to erlfmt_cli.
Now check also doesn't need to read the whole string, if it is just going to skip the file.
The format_string and format_file API functions are now even more symmetric.

File: src/erlfmt_cli.erl (hunk 5 above)
Comment (michalmuskala) [reply src/erlfmt_cli.erl:34]: `--verbose` doesn't use `-v` becuase it's already used for `--version`. We could use `-V` for alias to `--verbose`

Comment (awalterschulze) [reply src/erlfmt_cli.erl:34]: In general, I think it is okay to leave out the short hands for everything that isn't `-h`.
But I am okay with add the `-c` back, but `-V` seems weird.

Comment (awalterschulze) [reply src/erlfmt_cli.erl]: Added back -c

File: src/erlfmt_cli.erl [hunk 6]
```diff
@@ -84,38 +94,97 @@ do_unprotected(Opts, Name) ->
...
+        _ ->
+            ok
+    end,
+    case Result of
+        {ok, FormattedText, Warnings} ->
             [print_error_info(Warning) || Warning <- Warnings],
-            ok;
-        {check_failed, _, _, Warnings} ->
+            case Check of
+                true -> ok;
```
Comment (michalmuskala) [inline src/erlfmt_cli.erl:120]: Do we actually perform the check now? I can't see when it is done

**Review by michalmuskala — COMMENTED** [review]
I like this approach much more, but I think the `erlfmt_cli` code has a lot of spaghetti conditionals now and I find it rather hard to reason through. Especially the `check` flag is checked again and again. Perhaps there's some better way to do this?

File: src/erlfmt_cli.erl [hunk 7]
```diff
@@ -84,38 +94,97 @@ do_unprotected(Opts, Name) ->
...
             error
     end.
 
+write_formatted(_FileName, Formatted, standard_out) ->
+    io:put_chars(Formatted);
+write_formatted(FileName, Formatted, Out) ->
+    OutFileName = out_file(FileName, Out),
+    case filelib:ensure_dir(OutFileName) of
+        ok -> ok;
+        {error, Reason1} -> throw({error, {OutFileName, 0, file, Reason1}})
```
[the remark this answers is not in the document]
Comment (awalterschulze) [reply src/erlfmt_cli.erl]: I changed it to print and not throw

File: src/erlfmt_cli.erl (hunk 6 above)
Comment (awalterschulze) [reply src/erlfmt_cli.erl]: We performed check on stdin and not on files.
I fixed this now
And reran the test plan properly this time.
Good catch, thank you

**Comment by awalterschulze** [comment]
> I like this approach much more, great idea! I think the `erlfmt_cli` code, though, has a lot of spaghetti conditionals now and I find it rather hard to reason through. Especially the `check` flag is checked again and again. Perhaps there's some better way to do this?

I now made `check` part of `out()` again and so removed one level of nesting.

**Review by michalmuskala — APPROVED** [review]
